src/levenberg.jl

```diff
@@ -0,0 +1,408 @@
+    end
+end
+
+function jacobian_caches(alg::LevenbergMarquardt, f, u, p, ::Val{true})
+    uf = JacobianWrapper(f, p)
+    J = ArrayInterfaceCore.undefmatrix(u)
+
+    linprob = LinearProblem(J, _vec(zero(u)); u0 = _vec(zero(u)))
+    weight = similar(u)
+    recursivefill!(weight, false)
+
+    Pl, Pr = wrapprecs(alg.precs(J, nothing, u, p, nothing, nothing, nothing, nothing,
+                                 nothing)..., weight)
+    linsolve = init(linprob, alg.linsolve, alias_A = true, alias_b = true,
+                    Pl = Pl, Pr = Pr)
+
+    du1 = zero(u)
+    du2 = zero(u)
+    tmp = zero(u)
+    jac_config = build_jac_config(alg, f, uf, du1, u, tmp, du2)
+
+    uf, linsolve, J, du1, jac_config
+end
+
+function jacobian_caches(alg::LevenbergMarquardt, f, u, p, ::Val{false})
+    JacobianWrapper(f, p), nothing, ArrayInterfaceCore.undefmatrix(u), nothing, nothing
+end
+
+function SciMLBase.__init(prob::NonlinearProblem{uType, iip}, alg::LevenbergMarquardt,
+                          args...;
+                          alias_u0 = false,
+                          maxiters = 1000,
+                          abstol = 1e-6,
+                          internalnorm = DEFAULT_NORM,
+                          kwargs...) where {uType, iip}
+    if alias_u0
+        u = prob.u0
+    else
+        u = deepcopy(prob.u0)
+    end
+    f = prob.f
+    p = prob.p
+    if iip
+        fu = zero(u)
+        f(fu, u, p)
+    else
+        fu = f(u, p)
+    end
+    uf, linsolve, J, du_tmp, jac_config = jacobian_caches(alg, f, u, p, Val(iip))
+
+    if u isa Number
+        DᵀD = alg.min_damping_D
+    else
+        d = similar(u)
+        d .= alg.min_damping_D
+        DᵀD = Diagonal(d)
+    end
+
+    loss = internalnorm(fu)
+    JᵀJ = zero(J)
+    v = zero(u)
+    a = zero(u)
+    tmp_vec = zero(u)
+    v_old = zero(u)
+    δ = zero(u)
+    make_new_J = true
+    fu_tmp = zero(fu)
+
+    return LevenbergMarquardtCache{iip}(f, alg, u, fu, p, uf, linsolve, J, du_tmp,
+                                        jac_config,
+                                        1, false, maxiters, internalnorm,
+                                        ReturnCode.Default, abstol, prob, DᵀD, JᵀJ,
+                                        alg.damping_initial, alg.damping_increase_factor,
+                                        v, a, tmp_vec, v_old, loss, δ, loss, make_new_J,
+                                        fu_tmp)
+end
+function perform_step!(cache::LevenbergMarquardtCache{true})
+    @unpack fu, f, make_new_J = cache
+    if iszero(fu)
+        cache.force_stop = true
+        return nothing
+    end
+    if make_new_J
+        jacobian!(cache.J, cache)
+        mul!(cache.JᵀJ, cache.J', cache.J)
+        cache.DᵀD .= max.(cache.DᵀD, Diagonal(cache.JᵀJ))
+        cache.make_new_J = false
+    end
+    @unpack u, p, λ, JᵀJ, DᵀD, J = cache
+
+    # Usual Levenberg-Marquardt step ("velocity").
+    cache.v = -(JᵀJ .+ λ .* DᵀD) \ mul!(cache.du_tmp, J', fu)
```

Comment:
`-(JᵀJ .+ λ .* DᵀD)` should be put into a temporary

Reply:
`mul!(cache.du_tmp, J', fu)` I'd move that to a different line for clarity. Though it returns `cache.du_tmp` in most dispatches, it's usually best to mutate in one line and then use in the next